src/smackw32/smackw32.c

```diff
@@ -66,34 +247,46 @@ void SmackToBuffer(Smack* smack, uint32_t left, uint32_t top, uint32_t pitch, ui
 
     char* char_buf = buf;
 
-    const unsigned char* frame = smk_get_video(smack->smk_handle);
+    const unsigned char* frame = smk_get_video(user->smk_handle);
     for (i = 0; i < smack->Height; i++) {
         memcpy(&char_buf[(i * pitch)], &frame[i * smack->Width], smack->Width);
     }
 }
 
 uint32_t SmackDoFrame(Smack* smack) {
-    smack_last_frame_time = gHarness_platform.GetTicks();
+    SmackUserData *const user = smack->user;
+
+    user->smack_last_frame_time = gHarness_platform.GetTicks();
+
+    audio_step(smack);
 
-    // TODO: audio processing
     return 0;
 }
 
 void SmackNextFrame(Smack* smack) {
-    smk_next(smack->smk_handle);
+    SmackUserData *const user = smack->user;
+
+    smk_next(user->smk_handle);
     copy_palette(smack);
 }
 
 uint32_t SmackWait(Smack* smack) {
+    SmackUserData *const user = smack->user;
+
     uint32_t now = gHarness_platform.GetTicks();
-    if (now < smack_last_frame_time + smack->MSPerFrame) {
+    /* FIXME: get rid of bodge term */
+    if (now < user->smack_last_frame_time + smack->MSPerFrame - 5) {
```

Comment:
I did this to avoid stuttering.
When sleeping too long, the next audio frame is not ready soon enough.